src/core/manager.c

```diff
@@ -1663,7 +1663,11 @@ static void manager_invoke_notify_message(Manager *m, Unit *u, pid_t pid, const
         assert(m);
         assert(u);
         assert(buf);
-        assert(n > 0);
+
+        if(n <= 0) {
+            log_unit_debug(u, "Got a zero-length notification message. Ignoring");
+            return;
+        }
```

Comment:
Sorry, but the attitude "the caller must ensure they won't trip an assert() in the callee" is exactly what led to the state where any unprivileged user can take down pid1, precisely because manager_dispatch_notify_fd() should have but did not do that. So unless you've some other means of programatically verifying that it can't/won't happen again, you're leaving a time bomb waiting for the next refactoring that occurs.